File: yyevac/src/main/java/com/yy/yyeva/util/EvaMediaUtil.kt
package com.yy.yyeva.util

import android.media.MediaCodecList
import android.media.MediaExtractor
import android.media.MediaFormat
import com.yy.yyeva.file.IEvaFileContainer
import kotlin.collections.HashMap


object EvaMediaUtil {

    private const val TAG = "${EvaConstant.TAG}.MediaUtil"

    private var isTypeMapInit = false
    private val supportTypeMap = HashMap<String, Boolean>()

    const val MIME_HEVC = "video/hevc"

    fun getExtractor(evaFile: IEvaFileContainer): MediaExtractor {
        val extractor = MediaExtractor()
        evaFile.setDataSource(extractor)
        return extractor
    }

    /**
     * 是否为h265的视频
     */
    fun checkIsHevc(videoFormat: MediaFormat):Boolean {
        val mime = videoFormat.getString(MediaFormat.KEY_MIME) ?: ""
        return mime.contains("hevc")
    }

    fun selectVideoTrack(extractor: MediaExtractor): Int {
        val numTracks = extractor.trackCount
        for (i in 0 until numTracks) {
            val format = extractor.getTrackFormat(i)
            val mime = format.getString(MediaFormat.KEY_MIME) ?: ""
            if (mime.startsWith("video/")) {
                ELog.i(TAG, "Extractor selected track $i ($mime): $format")
                return i
            }
        }
        return -1
    }

    fun selectAudioTrack(extractor: MediaExtractor): Int {
        val numTracks = extractor.trackCount
        for (i in 0 until numTracks) {
            val format = extractor.getTrackFormat(i)
            val mime = format.getString(MediaFormat.KEY_MIME) ?: ""
            if (mime.startsWith("audio/")) {
                ELog.i(TAG, "Extractor selected track $i ($mime): $format")
                return i
            }
        }
        return -1
    }

    /**
     * 检查设备解码支持类型
     */
    @Synchronized
    fun checkSupportCodec(mimeType: String): Boolean {
        if (!isTypeMapInit) {
            isTypeMapInit = true
            getSupportType()
        }
        return supportTypeMap.containsKey(mimeType.toLowerCase())
    }


    private fun getSupportType() {
        try {
            val numCodecs = MediaCodecList.getCodecCount()
            for (i in 0 until numCodecs) {
                val codecInfo = MediaCodecList.getCodecInfoAt(i)
                if (!codecInfo.isEncoder) {
                    continue
                }
                val types = codecInfo.supportedTypes
                for (j in types.indices) {
                    supportTypeMap[types[j].toLowerCase()] = true
                }
            }
            ELog.i(TAG, "supportType=${supportTypeMap.keys}")
        } catch (t: Throwable) {
            ELog.e(TAG, "getSupportType $t")
        }
    }

}

File: yyevac/src/main/cpp/util/evamediautil.cpp
//
// Created by asus on 2022/6/25.
//

#include "evamediautil.h"

EvaMediaUtil* EvaMediaUtil::instance = nullptr;

EvaMediaUtil::~EvaMediaUtil() {
    if (EvaMediaUtil::instance) {
        delete EvaMediaUtil::instance;
        EvaMediaUtil::instance = nullptr;
    }
}

AMediaExtractor *EvaMediaUtil::getExtractor(IEvaFileContainer *evaFile) {
    AMediaExtractor* extractor = AMediaExtractor_new();
    evaFile->setDataSource(extractor);
    return extractor;
}

bool EvaMediaUtil::checkIsHevc(AMediaFormat *videoFormat) {
    const char* mime;
    AMediaFormat_getString(videoFormat, AMEDIAFORMAT_KEY_MIME, &mime);
    return false;
}

int EvaMediaUtil::selectVideoTrack(AMediaExtractor *extractor) {
    int numTracks = AMediaExtractor_getTrackCount(extractor);
    string head("video/");
    for (int i = 0; i < numTracks; ++i) {
        AMediaFormat* format = AMediaExtractor_getTrackFormat(extractor, i);
        const char *mime;

        if(!AMediaFormat_getString(format,AMEDIAFORMAT_KEY_MIME, &mime)) {
            ELOGE("selectVideoTrack getMime error");
        }
        if (mime != nullptr) {
            string m (mime);
            if (startWith(m, head)) {
                ELOGV("Extractor selected track %i, %s", i, m.c_str());
                return i;
            }
        }
    }
    return -1;
}

bool EvaMediaUtil::startWith(string src, string start) {
    if (src == "" || start == "" || src.size() < start.size()) {
        return false;
    }

    return src.compare(0, start.size(), start) == 0;
}

int EvaMediaUtil::selectAudioTrack(AMediaExtractor *extractor) {
    int numTracks = AMediaExtractor_getTrackCount(extractor);
    string head("audio/");
    for (int i = 0; i < numTracks; ++i) {
        AMediaFormat* format = AMediaExtractor_getTrackFormat(extractor, i);
        const char*  mime;
        if(!AMediaFormat_getString(format, AMEDIAFORMAT_KEY_MIME, &mime)) {
            ELOGE("selectAudioTrack getMime error %i",i);
            continue;
        }
        if (mime != nullptr) {
            string m (mime);
            if (startWith(m, head)) {
                ELOGV("Extractor selected track %i, %s", i, m.c_str());
                return i;
            }
        }
    }
    return -1;
}

bool EvaMediaUtil::checkSupportCodec(string mimeType) {
    if (!isTypeMapInit) {
        isTypeMapInit = true;
        getSupportType();
    }
    //判断key在map中出现的次数
    return supportTypeMap.count(mimeType) != 0;
}

void EvaMediaUtil::getSupportType() {
    try {

    } catch (exception e) {
      ELOGE("getSupportType error %s",e.what());
    }
}

//EvaMediaUtil *EvaMediaUtil::getInstance() {
//    if (instance == nullptr) {
//        instance = new EvaMediaUtil();
//    }
//    return instance;
//}
